snap/epoch.go

```diff
@@ -0,0 +1,288 @@
+// write attribute defaults to the last item in the read attribute. If both are
+// unset, it's the same as not specifying an epoch at all (i.e. epoch: 0). The
+// lists must be in ascending order, and there must be a non-empty intersection
+// between them.
+//
+// Epoch numbers must be written as base 10 numbers, with no zero padding.
+type Epoch struct {
+	Read  []uint32 `yaml:"read"`
+	Write []uint32 `yaml:"write"`
+}
+
+// E returns the epoch represented by the expression s. It's meant for use in
+// testing, as it panics at the first sign of trouble.
+func E(s string) Epoch {
+	var e Epoch
+	if err := e.fromEpochString(s); err != nil {
+		panic(fmt.Errorf("%q: %v", s, err))
+	}
+	return e
+}
+
+// EpochZero returns the epoch represented by the expression "0".
+func EpochZero() Epoch {
+	return Epoch{Read: []uint32{0}, Write: []uint32{0}}
+}
+
+// IsNull checks whether this epoch looks like it's not been initialized.
+func (e Epoch) IsNull() bool {
+	return len(e.Read) == 0 && len(e.Write) == 0
+}
+
+func (e *Epoch) fromEpochString(s string) error {
+	if len(s) == 0 {
+		e.Read = []uint32{0}
+		e.Write = []uint32{0}
+		return nil
+	}
+	splat := false
+	if s[len(s)-1] == '*' {
+		splat = true
+		s = s[:len(s)-1]
+	}
+	n, err := parseInt(s)
+	if err != nil {
+		return err
+	}
+	if splat {
+		if n == 0 {
+			return ErrEpochOhSplat
+		}
+		e.Read = []uint32{n - 1, n}
+	} else {
+		e.Read = []uint32{n}
+	}
+	e.Write = []uint32{n}
+
+	return nil
+}
+
+func (e *Epoch) fromStructured(newEpoch structuredEpoch) error {
+	if newEpoch.Read == nil {
+		if newEpoch.Write == nil {
+			newEpoch.Write = []uint32{0}
+		}
+		newEpoch.Read = newEpoch.Write
+	} else if len(newEpoch.Read) == 0 {
+		// this means they explicitly set it to []. Bad they!
+		return ErrEmptyEpochList
+	}
+	if newEpoch.Write == nil {
+		newEpoch.Write = newEpoch.Read[len(newEpoch.Read)-1:]
+	} else if len(newEpoch.Write) == 0 {
+		return ErrEmptyEpochList
+	}
+	e.Read = newEpoch.Read
+	e.Write = newEpoch.Write
+
+	if err := e.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *Epoch) UnmarshalJSON(bs []byte) error {
+	return e.UnmarshalYAML(func(v interface{}) error {
+		return json.Unmarshal(bs, &v)
+	})
+}
+
+func (e *Epoch) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var oldEpoch string
+	if err := unmarshal(&oldEpoch); err == nil {
+		return e.fromEpochString(oldEpoch)
+	}
+	var newEpoch structuredEpoch
+	if err := unmarshal(&newEpoch); err != nil {
+		return err
+	}
+
+	return e.fromStructured(newEpoch)
+}
+
+// Validate checks that the epoch makes sense.
+func (e *Epoch) Validate() error {
+	if e == nil || len(e.Read) == 0 || len(e.Write) == 0 {
+		return ErrEmptyEpochList
+	}
+	if !(sort.IsSorted(uint32Slice(e.Read)) && sort.IsSorted(uint32Slice(e.Write))) {
+		return ErrEpochListNotSorted
+	}
+	foundOne := false
+	for i := range e.Read {
+		idx := sort.Search(len(e.Write), func(j int) bool { return e.Write[j] >= e.Read[i] })
```

Comment:
as it's user-originating data I went for the smaller-O function, but if we cap it at 10 (or even 100), +1 to the trivial double loop.

Reply:
capping it at 10, doing the trivial

Reply:
From the code now:
```golang
	// O(𝑚𝑛) instead of O(𝑚log𝑛) for the binary search we could do, but
	// 𝑚 and 𝑛 <10 per above, so the simple solution is good enough (and if
	// that alone makes you nervous, know that it is ~2× faster in the
	// worst case; bisect starts being faster at ~50 entries).
```